src/hash.c

```diff
@@ -288,19 +369,36 @@ static inline void raw_hash_fill(uint32_t* p, R_len_t size, SEXP x) {
     p[i] = hash_combine(p[i], HASHER(x, i));    \
   }
 
-static void list_hash_fill(uint32_t* p, R_len_t size, SEXP x) {
-  HASH_FILL_BARRIER(list_hash_scalar);
+#define HASH_FILL_BARRIER_NA_PROPAGATE(HASHER)  \
+  for (R_len_t i = 0; i < size; ++i) {          \
+    if (p[i] == HASH_MISSING) {                 \
+      continue;                                 \
+    }                                           \
+    uint32_t hash = HASHER(x, i);               \
+    if (hash == HASH_MISSING) {                 \
+      p[i] = HASH_MISSING;                      \
+    } else {                                    \
+      p[i] = hash_combine(p[i], HASHER(x, i));  \
+    }                                           \
```

Comment:
Calling `HASHER(x, i)` twice here, can reuse `hash`. Maybe call this `elt_hash` if you also store `uint32_t hash = p[i]`?

Reply:
oops